numba/cuda/tests/cudapy/test_lineinfo.py

```diff
@@ -1,42 +1,207 @@
+        # signal an exception (e.g. divide by zero) has occurred. When the
+        # error model is the default NumPy one (as it should be when only
+        # lineinfo is enabled) the device function always returns 0.
+        self.assertNotIn('ret i32 1', llvm)
+
+    def test_no_lineinfo_in_device_function(self):
+        # Ensure that no lineinfo is generated in device functions by default.
+        @cuda.jit
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit
+        def caller(x):
+            x[0] = 1
+            callee(x)
+
+        sig = (int32[:],)
+        self._check(caller, sig=sig, expect=False)
+
+    def test_lineinfo_in_device_function(self):
+        if not NVVM().is_nvvm70:
+            self.skipTest("lineinfo not generated for NVVM 3.4")
+
+        # First we define a device function / kernel pair and run the usual
+        # checks on the generated LLVM and PTX.
+
+        @cuda.jit(lineinfo=True)
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit(lineinfo=True)
+        def caller(x):
+            x[0] = 1
+            callee(x)
+
+        sig = (int32[:],)
+        self._check(caller, sig=sig, expect=True)
+
+        # Now we can check the PTX of the device function specifically.
+
+        ptx = caller.inspect_asm(sig)
+        ptxlines = ptx.splitlines()
+
+        # To check the device function, we need to identify its boundaries.
+
+        # A line beginning with ".weak .func"
+        devfn_start = re.compile(r'^\.weak\s*\.func')
+
+        # Identify the beginning of the function.
+        start = None
+
+        for lineno, line in enumerate(ptxlines):
+            if devfn_start.match(line) is not None:
+                # We will begin our search on the line following the
+                # declaration
+                start = lineno + 1
+
+        if start is None:
+            self.fail(f'Could not identify device function in:\n\n{ptx}')
+
+        # Identify the end of the function
+        end = None
+
+        for offset, line in enumerate(ptxlines[start:]):
+            # Assume the end of the function is a line with an unindented '}'
+            if line[:1] == '}':
+                end = start + offset
+
+        if end is None:
+            self.fail(f'Could not identify end of device function in:\n\n{ptx}')
+
+        # Scan for .loc directives in the device function.
+        loc_directive = self._loc_directive_regex()
+        found = False
+
+        for line in ptxlines[start:end]:
+            if loc_directive.search(line) is not None:
+                found = True
+
+        if not found:
+            # Join one line either side so the function as a whole is shown,
+            # i.e. including the declaration and parameter list, and the
+            # closing brace.
+            devfn = "\n".join(ptxlines[start - 1:end + 1])
+            self.fail(f'.loc directive not found in:\n\n{devfn}')
```

Comment:
```suggestion
        for line in ptxlines[start:end]:
            if loc_directive.search(line) is not None:
                break
        else:
            # Join one line either side so the function as a whole is shown,
            # i.e. including the declaration and parameter list, and the
            # closing brace.
            devfn = "\n".join(ptxlines[start - 1:end + 1])
            self.fail(f'.loc directive not found in:\n\n{devfn}')
```
suggest using `for...else...`

Reply:
I'm not a big fan of `for .. else ..` because it makes the condition ("found") implicit rather than explicit, and therefore harder for a reader to understand. Do you feel strongly the other way?

Reply:
I don't feel strongly either way, but would prefer explicit `break` on the condition being found. I made the suggestion after noticing there was no `break` on being found, then realised it was a `for...else..`,

Reply:
I've now removed this code anyway (it's not needed when we're generating code "correctly" without performance regression with lineinfo).